### PR TITLE
usb: stm32: Introduce transfer method

### PR DESCRIPTION
**Pull request by loicpoulain**
Transfer is an important concept of USB specification.
During a transfer, several packets can be transmitted.
Today the usb API only provides ep_read and ep_write
functions but the transfer concept is missing.

This is typically ok for basic drivers which don't have
to take care of how data is transfered/splitted.
However there are some drivers like CDC EEM, in which
transfer concept is important for packet management.

Moreover, current ep-write and ep_read method have a
different implementation in usb_dc_dw and usb_dc_stm32
device drivers. For example usb_dc_dw supports only
1-data-packet transfers due to its implementation.
This can increase latency and reduce performance.

I think this is something we need to fix/improve by
implementing better transfer management.

This patch introduces usb_dc_ ep_transfer method which
can be used to configure IN/OUT transfers. This allows
to configure and request different transfer sizes and
should prevent usage of the current stm32 temporary
buffer.

This method has asynchronous and synchronous mode.
Synchronous mode waits for transfer completion before
returning. Asynchronouse mode (irq safe) configures
the transfer and returns immediately, the provided
callback will be then called on transfer completion.

This also update ep_write and ep_read stm32 implementation
to use this new method but keep their behavior unchanged
for legacy reasons.

Note that for now this method is local to stm32 device
driver, however the goal would be to expose this function
as a new USB device driver API method so that class
drivers use it. This will request same implementation in
the usb_dc_dw_driver.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

### CODE REVIEW

**Comment by loicpoulain** [comment]
@ydamigos, @jfischer-phytec-iot, @finikorg could you please have a look. In a next step, I would like to implement the same for usb_dc_dw and then make the usb_dc_transfer part of the API and so available for class drivers. Having such function makes sense to me and should simplify some drivers, but let me thoughts...

File: drivers/usb/device/usb_dc_stm32.c [hunk 1]
```diff
@@ -243,10 +252,105 @@ int usb_dc_set_address(const u8_t addr)
 	return 0;
 }
 
-int usb_dc_ep_start_read(u8_t ep, u8_t *data, u32_t max_data_len)
+static int usb_dc_ep_transfer(const u8_t ep, u8_t *buf, size_t dlen,
+			      bool is_in, usb_dc_transfer_callback cb)
```
Comment (ydamigos) [inline drivers/usb/device/usb_dc_stm32.c]: Please align.

**Review by ydamigos — COMMENTED** [review]
Does this also support read_wait/conitinue API (PR #5166)?

File: drivers/usb/device/usb_dc_stm32.c (hunk 1 above)
Comment (loicpoulain) [reply drivers/usb/device/usb_dc_stm32.c]: seems to be a github formatting issue here

**Comment by loicpoulain** [comment]
Not support read_wait/continue but I can definitely take the patch in this PR and rebase.

**Review by ydamigos — APPROVED** [review]
LGTM

**Comment by loicpoulain** [comment]
@galak, could you please check this PR

**Comment by loicpoulain** [comment]
@galak could we move forward and merge this PR ?

**Comment by jfischer-phytec-iot** [comment]
@loicpoulain Where is the transfer concept? It seems to be tailored to the STM HAL Driver. I do not see how it can be exposed to the API. BTW, everything initiated by the host is running in the ISR context.

**Comment by loicpoulain** [comment]
@jfischer-phytec-iot, Yes in case of STM32, the 'transfer' is mostly handled at HAL level (manage tx/rx buffer, fill the fifo when empty...), the driver here mainly manages transfer state, concurrency and completion. We can see this as a (very)simplified version of USB URB (on Linux).

**Comment by loicpoulain** [comment]
@ydamigos, can we merge this PR?

**Comment by galak** [comment]
Added @finikorg to see if he can do a review

**Comment by finikorg** [comment]
@galak Sorry, haven't had time to review it, I will check this and DW code next week.

**Review by finikorg — APPROVED** [review]
Looks OK

**Comment by galak** [comment]
@loicpoulain can you look to rebase this, looks like its gotten a little out of date now.

**Comment by loicpoulain** [comment]
@galak , done!

**Pull request merged**